liquibase/liquibase
# Improve nested changelog precondition handling

Fixes #1624 by added an addtional GlobalPreconditionContainer level outside the generic PreconditionContainer

- - -
## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: [https://github.com/liquibase/liquibase/issues/1624](https://github.com/liquibase/liquibase/issues/1624)
* Local Branch: [https://github.com/liquibase/liquibase/tree/handle-nested-preconditions-with-onsqloutput](https://github.com/liquibase/liquibase/tree/handle-nested-preconditions-with-onsqloutput)
* Unit Tests: [https://github.com/liquibase/liquibase/pull/2020/checks](https://github.com/liquibase/liquibase/pull/2020/checks)
* Integration Tests: (shown in Unit Tests link above)
* Functional Tests: [https://jenkins.datical.net/job/liquibase-pro/job/handle-nested-preconditions-with-onsqloutput/](https://jenkins.datical.net/job/liquibase-pro/job/handle-nested-preconditions-with-onsqloutput/)

#### Testing
* Steps to Reproduce: [https://github.com/liquibase/liquibase/issues/1624](https://github.com/liquibase/liquibase/issues/1624) especially the examples in the comments.
* Guidance:
  * Did not have a great way to add automated tests for checking the issue

#### Dev Verification
Manually tested variations of:

* Having a root precondition block or not
* Setting onSqlOutput to various settings on the root changelog and child changelogs to see that they work as expected

## Test Requirements (Internal Liquibase QA)
**Manual Tests**

The bug here is that when there is a precondition in a child changelog but no corresponding precondition in the parent changelog, the precondition in the child changelog is not executed during update-sql. A changelog is considered a “child” if it is included as a file in the “parent” changelog.  When this bug is fixed, the child precondition will execute as expected when there is no precondition in the parent changelog.

[https://b-gyula.github.io/liquibase-doc/documentation/preconditions](https://b-gyula.github.io/liquibase-doc/documentation/preconditions|smart-link) 

This bug and fix are database agnostic. The parent and child changelogs attached to this ticket should work for any of our five Liquibase certified database platforms; I did my initial repro of the bug using MySQL. You can use whatever database makes you happiest. :slightly_smiling_face:  The changelogs are the starting point for the tests; you will need to update the preconditions to evaluate the different scenarios.

_Verify update-sql executes a onSqlOutput=TEST precondition for a child changelog where the condition is not met._

* Ensure the parent changelog does **not** contain a precondition
* Ensure the child changelog has a tableExists precondition with onSqlOutput=TEST
* Ensure the child changelog precondition specifies a tableName of a table that is **not** present on the database.
* Validate that update-sql executes the precondition 
  * The command should not generate any SQL
  * The command should output that a precondition failed because the condition was not met.

_Verify update-sql executes a onSqlOutput=TEST precondition in a child changelog where the condition is met._

* Ensure the parent changelog does **not** contain a precondition
* Ensure the child changelog has a tableExists precondition with onSqlOutput=TEST
* Ensure the child changelog precondition specifies a tableName of a table that **is** present on the database.
* Validate that update-sql executes the precondition 
  * The command outputs SQL because the precondition is met.

_Verify update-sql executes a onSqlOutput=FAIL precondition in a child changelog and the condition is met._

* Ensure the parent changelog does **not** contain a precondition
* Ensure the child changelog has a tableExists precondition with onSqlOutput=FAIL
* Ensure the child changelog precondition specifies a tableName of a table that **is** present on the database.
* Validate that update-sql executes the precondition 
  * The command should not generate any SQL
  * The command should output that a precondition failed because the condition was not met. (Check with developer to find out what the console output to expect in this case).

_Verify update-sql does not execute an onSqlOutput=IGNORE precondition in a child changelog._

* Ensure the parent changelog does **not** contain a precondition
* Ensure the child changelog has a tableExists precondition with onSqlOutput=IGNORE
* Ensure the child changelog precondition specifies a tableName of a table that **is not** present on the database.
* Validate that update-sql executes the precondition 
  * The command should generate SQL even though the condition is not met.

_Verify update-sql executes a onSqlOutput=TEST precondition in a child changelog when an empty TEST precondition is in the parent changelog._

* Ensure the parent changelog does has a precondition defined without a condition. For example:
`<preConditions onSqlOutput="TEST"/>`
* Ensure the child changelog has a tableExists precondition with onSqlOutput=TEST
* Ensure the child changelog precondition specifies a tableName of a table that is **not** present on the database.
* Validate that update-sql executes the precondition 
  * The command should not generate any SQL
  * The command should output that a precondition failed because the condition was not met.

**Automated Tests**

* The developer test frameworks did not have a good place to add tests; we will add a functional test.
* Add a new functional test to [tests/functional/update/PreconditionsUpdateTests.java](https://github.com/liquibase/liquibase-pro-tests/blob/master/src/test/java/tests/functional/update/PreconditionsUpdateTests.java) This fix is not database-specific. I suggest using MariaDB OR MySQL OR Postgres, as all of those run fairly quickly in builds.
  * Verify changelog-level preconditions apply to included (child) changelogs.
  * GIVEN two changelogs (parent and child)
  * AND the parent changelog includes the child
  * AND the child changelog has a changelog-level precondition to execute TEST onSqlOutput
  * AND the child changelog precondition is testing for the existence of a table
  * AND the expected table does not exist in the changelog or on the database
  * WHEN I run update-sql
  * THEN I expect update-sql to halt execution due to the unmet precondition
  * AND output to contain:
  * ```
liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Validation Failed:
     1 preconditions failed
     LB2104-parent-changelog.xml : Table testdb1.tbl_does_not_exist does not exist
```



#### Github Users
* @dewa-actico



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-2104) by [Unito](https://www.unito.io)
